#5861 Changes query which improves performance

eKuG opened the pull request:
### Summary

The earlier query had several performance bottlenecks, which results in slow execution times as the dataset size increases. Below is a summary of the bottlenecks and proposed strategies for optimization.

### Bottlenecks Identified

1. **Distinct on (traceID)**:
    - The `DISTINCT ON (traceID)` adds significant overhead to the query execution due to the de-duplication process on the large dataset.
2. **Global Inner Join on the Same Table**:
    - Performing a `GLOBAL INNER JOIN` on the same table (`signoz_index_v2`) results in increased computational complexity and disk I/O as Clickhouse handles large-scale joins across a distributed cluster.
3. **No Indexes on `parentSpanID` and `traceID`**:
    - **Although indexes have been added**, they have not led to the expected level of optimization. This suggests that the query is not leveraging the indexes effectively or additional tuning is required.
    - Add index
      - `ALTER TABLE signoz_traces.signoz_index_v2 ADD INDEX idx_traceID traceID TYPE bloom_filter GRANULARITY 4;`
      - `ALTER TABLE signoz_traces.signoz_index_v2 ADD INDEX idx_parentSpanID parentSpanID TYPE bloom_filter GRANULARITY 4;`
      - `ALTER TABLE signoz_traces.signoz_index_v2 ADD INDEX idx_parentSpanID_timestamp (parentSpanID, timestamp) TYPE minmax GRANULARITY 1;`

#### Related Issues / PR's

[#5861](https://github.com/SigNoz/signoz/issues/5861)

**More details regarding this PR has been documented and can be found on Notion**

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SigNoz/signoz?pullRequest=6081) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SigNoz/signoz?pullRequest=6081) before we can accept your contribution.<br/><hr/>**Ekansh Gupta** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SigNoz/signoz?pullRequest=6081) it.</sub>

eKuG commented:
@nityanandagohain checks have paased

nityanandagohain commented on pkg/query-service/app/traces/v3/query_builder_test.go:
```diff
@@ -1162,13 +1162,13 @@ var testBuildTracesQueryData = []struct {
+		"FROM signoz_traces.distributed_signoz_index_v2 WHERE parentSpanID = '' AND (timestamp >= '1680066360726210000' AND timestamp <= '1680066458000000000')  " + 
+		"AND stringTagMap['method'] = 'GET' ORDER BY durationNano DESC  LIMIT 100)" +
+		" AS inner_subquery ) AS subQuery " + 
+		"ON signoz_traces.distributed_signoz_index_v2.traceID = subQuery.traceID GROUP BY subQuery.traceID, subQuery.durationNano, subQuery.name, subQuery.serviceName ORDER BY subQuery.durationNano desc;",
```
according to the example you shared, there should be a timestamp filter in the outer query ?

nityanandagohain commented on pkg/query-service/app/traces/v3/query_builder.go:
```diff
@@ -363,7 +363,8 @@ func buildTracesQuery(start, end, step int64, mq *v3.BuilderQuery, _ string, pan
 			}
-			query = withSubQuery + ") " + fmt.Sprintf(constants.TracesExplorerViewSQLSelectQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_TABLENAME, constants.SIGNOZ_SPAN_INDEX_TABLENAME)
+			// query = withSubQuery + ") " + fmt.Sprintf(constants.TracesExplorerViewSQLSelectQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_TABLENAME, constants.SIGNOZ_SPAN_INDEX_TABLENAME)
+			query = fmt.Sprintf(constants.TracesExplorerViewSQLSelectBeforeSubQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_TABLENAME) + withSubQuery + ") " + fmt.Sprintf(constants.TracesExplorerViewSQLSelectAfterSubQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_TABLENAME)
```
nit: instead of before, after string constants a single constant string with one fmt.Sprintf will be cleaner?

eKuG replied:
It would be cleeaner, I'm just making the whole thing more readable. Ensures debuggability if needed.

eKuG commented:
@srikanthccv waiting for your review 😅

srikanthccv commented:
>Performing a GLOBAL INNER JOIN on the same table (signoz_index_v2) results in increased computational complexity and disk I/O as Clickhouse handles large-scale joins across a distributed cluster.

Just to make sure I follow the point, we are not joining at the same table, right? We are joining the result table of the subquery.

Also, I guess the improvement is coming from the existence of `WHERE timestamp >= '1726637898000000000' AND timestamp <= '1726724298000000000' ` on the new query. (optional) Can you add the same where clause (on the main query) filter to the existing query and share the results? I just wanted to understand where the gains are coming from because I think the queries are the same but have different flavours.

> GLOBAL INNER JOIN

The sharding key of table `signoz_traces.distributed_signoz_index_v2` is `cityHash64(traceID)`. This means all spans of the same trace will reside in the same shard. Given, this information you don't need  `GLOBAL INNER JOIN` it just needs to be `INNER JOIN`.

eKuG commented:
```
WITH subQuery AS
    (
        SELECT
            traceID,
            durationNano,
            serviceName,
            name
        FROM signoz_traces.distributed_signoz_index_v2
        WHERE (parentSpanID = '') AND (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
        ORDER BY durationNano DESC
        LIMIT 1 BY traceID
        LIMIT 10
    )
SELECT
    subQuery.serviceName,
    subQuery.name,
    COUNT() AS span_count,
    subQuery.durationNano,
    traceID
FROM signoz_traces.distributed_signoz_index_v2
GLOBAL INNER JOIN subQuery ON distributed_signoz_index_v2.traceID = subQuery.traceID
WHERE (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
GROUP BY
    traceID,
    subQuery.durationNano,
    subQuery.name,
    subQuery.serviceName
ORDER BY subQuery.durationNano DESC
SETTINGS max_threads = 12, min_bytes_to_use_direct_io = 1
```

> 10 rows in set. Elapsed: 19.284 sec. Processed 230.18 million rows, 8.66 GB (11.94 million rows/s., 449.03 MB/s.)
Peak memory usage: 1.73 GiB.

@srikanthccv the number of records processed is definitely lesser but the Peak memory usage and Latency hasn't improved.

eKuG commented:
> The sharding key of table signoz_traces.distributed_signoz_index_v2 is cityHash64(traceID). This means all spans of the same trace will reside in the same shard. Given, this information you don't need GLOBAL INNER JOIN it just needs to be INNER JOIN.

> Just to make sure I follow the point, we are not joining at the same table, right? We are joining the result table of the subquery.

**The answer to this question is YES**

srikanthccv commented:
is there a difference between in response time `GLOBAL INNER JOIN` and `INNER JOIN` (theoretically should be non-significant)?

srikanthccv commented:
Please share the comparison of the following.

local table join because same trace id should exist in same shard

```
SELECT
    subQuery.serviceName,
    subQuery.name,
    COUNT() AS span_count,
    subQuery.durationNano,
    subQuery.traceID
FROM signoz_traces.distributed_signoz_index_v2
INNER JOIN
(
    SELECT *
    FROM
    (
        SELECT
            traceID,
            durationNano,
            serviceName,
            name
        FROM signoz_traces.signoz_index_v2
        WHERE (parentSpanID = '') AND (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
        ORDER BY durationNano DESC
        LIMIT 1 BY traceID
        LIMIT 10
    ) AS inner_subquery
) AS subQuery ON signoz_traces.distributed_signoz_index_v2.traceID = subQuery.traceID
WHERE timestamp >= '1726637898000000000' AND timestamp <= '1726724298000000000'
GROUP BY
    subQuery.traceID,
    subQuery.durationNano,
    subQuery.name,
    subQuery.serviceName
ORDER BY subQuery.durationNano DESC;
```

global join

```
SELECT
    subQuery.serviceName,
    subQuery.name,
    COUNT() AS span_count,
    subQuery.durationNano,
    subQuery.traceID
FROM signoz_traces.distributed_signoz_index_v2
GLOBAL INNER JOIN
(
    SELECT *
    FROM
    (
        SELECT
            traceID,
            durationNano,
            serviceName,
            name
        FROM signoz_traces.distributed_signoz_index_v2
        WHERE (parentSpanID = '') AND (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
        ORDER BY durationNano DESC
        LIMIT 1 BY traceID
        LIMIT 10
    ) AS inner_subquery
) AS subQuery ON signoz_traces.distributed_signoz_index_v2.traceID = subQuery.traceID
WHERE timestamp >= '1726637898000000000' AND timestamp <= '1726724298000000000'
GROUP BY
    subQuery.traceID,
    subQuery.durationNano,
    subQuery.name,
    subQuery.serviceName
ORDER BY subQuery.durationNano DESC;
```

eKuG commented:
```
SELECT
    subQuery.serviceName,
    subQuery.name,
    COUNT() AS span_count,
    subQuery.durationNano,
    subQuery.traceID
FROM signoz_traces.distributed_signoz_index_v2
INNER JOIN
(
    SELECT *
    FROM
    (
        SELECT
            traceID,
            durationNano,
            serviceName,
            name
        FROM signoz_traces.distributed_signoz_index_v2
        WHERE (parentSpanID = '') AND (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
        ORDER BY durationNano DESC
        LIMIT 1 BY traceID
        LIMIT 10
    ) AS inner_subquery
) AS subQuery ON signoz_traces.distributed_signoz_index_v2.traceID = subQuery.traceID
WHERE (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
GROUP BY
    subQuery.traceID,
    subQuery.durationNano,
    subQuery.name,
    subQuery.serviceName
ORDER BY subQuery.durationNano DESC
SETTINGS max_threads = 12, min_bytes_to_use_direct_io = 1
```

> 10 rows in set. Elapsed: 13.940 sec. Processed 158.63 million rows, 6.07 GB (11.38 million rows/s., 435.53 MB/s.)
Peak memory usage: 935.99 MiB.

Performance gain is almost null @srikanthccv

srikanthccv commented:
@eKuG it should have given an error but I guess you only have one shard which explains why it didn't throw error or didn't show any improvement. The query I was suggesting to compare uses the local table instead of global table inside the sub query, like 

```sql
SELECT
    subQuery.serviceName,
    subQuery.name,
    COUNT() AS span_count,
    subQuery.durationNano,
    subQuery.traceID
FROM signoz_traces.distributed_signoz_index_v2
INNER JOIN
(
    SELECT *
    FROM
    (
        SELECT
            traceID,
            durationNano,
            serviceName,
            name
        FROM signoz_traces.signoz_index_v2
        WHERE (parentSpanID = '') AND (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
        ORDER BY durationNano DESC
        LIMIT 1 BY traceID
        LIMIT 10
    ) AS inner_subquery
) AS subQuery ON signoz_traces.distributed_signoz_index_v2.traceID = subQuery.traceID
WHERE (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
GROUP BY
    subQuery.traceID,
    subQuery.durationNano,
    subQuery.name,
    subQuery.serviceName
ORDER BY subQuery.durationNano DESC
LIMIT 1 BY subQuery.traceID
LIMIT 10
```

If there is only one shard both queries run the same but if there are 2 or more shards then each shard independently does the work. See this section for more details https://clickhouse.com/docs/en/sql-reference/operators/in#distributed-subqueries

eKuG commented:
```
SELECT
    subQuery.serviceName,
    subQuery.name,
    COUNT() AS span_count,
    subQuery.durationNano,
    subQuery.traceID
FROM signoz_traces.distributed_signoz_index_v2
INNER JOIN
(
    SELECT *
    FROM
    (
        SELECT
            traceID,
            durationNano,
            serviceName,
            name
        FROM signoz_traces.signoz_index_v2
        WHERE (parentSpanID = '') AND (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
        ORDER BY durationNano DESC
        LIMIT 1 BY traceID
        LIMIT 10
    ) AS inner_subquery
) AS subQuery ON signoz_traces.distributed_signoz_index_v2.traceID = subQuery.traceID
WHERE (timestamp >= '1726637898000000000') AND (timestamp <= '1726724298000000000')
GROUP BY
    subQuery.traceID,
    subQuery.durationNano,
    subQuery.name,
    subQuery.serviceName
ORDER BY subQuery.durationNano DESC
LIMIT 1 BY subQuery.traceID
LIMIT 10
SETTINGS max_threads = 12, min_bytes_to_use_direct_io = 1
```

> 10 rows in set. Elapsed: 13.950 sec. Processed 158.63 million rows, 6.07 GB (11.37 million rows/s., 435.20 MB/s.)
Peak memory usage: 925.68 MiB.

Understood your point @srikanthccv . 
I have also attached results for the query you mentioned.

srikanthccv commented:
Are you running on two shards or one shard? The suggested query won't show any difference if only one shard exists but can be effective when there are multiple shards.

eKuG commented:
Yeh @srikanthccv currenlty on the staging environment where I'm testing, I'm using 1 shard only. You're right about the observation.
Let me update the query in the pull request and get back to you.

<img width="314" alt="Screenshot 2024-10-01 at 3 26 14 PM" src="https://github.com/user-attachments/assets/28286a39-2a4c-4856-bc17-fbb90ab8c691">

ankitnayan commented:
@eKuG when do we plan to release this?

eKuG commented:
Hey @ankitnayan 
I have addressed all the comments which were raised by @srikanthccv . As soon as he approves the PR, we are ready to go. 
Earlier @nityanandagohain had approved this, but Srikant wanted some changes. 
So I'm just waiting for @srikanthccv on this.

ankitnayan commented:
> Hey @ankitnayan I have addressed all the comments which were raised by @srikanthccv . As soon as he approves the PR, we are ready to go. Earlier @nityanandagohain had approved this, but Srikant wanted some changes. So I'm just waiting for @srikanthccv on this.

You could ask/remind if PRs are waiting for a week.
